plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServer.java

```diff
@@ -0,0 +1,213 @@
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.NullProgressMonitor;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.debug.core.ILaunchConfigurationWorkingCopy;
+import org.eclipse.jst.server.core.IEnterpriseApplication;
+import org.eclipse.jst.server.core.IWebModule;
+import org.eclipse.wst.server.core.IModule;
+import org.eclipse.wst.server.core.IModuleType;
+import org.eclipse.wst.server.core.IServer;
+import org.eclipse.wst.server.core.ServerUtil;
+import org.eclipse.wst.server.core.internal.facets.FacetUtil;
+import org.eclipse.wst.server.core.model.ServerDelegate;
+
+import com.google.cloud.tools.eclipse.appengine.localserver.Activator;
+
+/**
+ * A {@link ServerDelegate} for Google Cloud SDK.
+ */
+@SuppressWarnings("restriction") // For FacetUtil
+public class CloudSdkServer extends ServerDelegate {
+  // must match the value used in plugin.xml in org.eclipse.wst.server.core.serverTypes
+  public static final String SERVER_TYPE_ID = "com.google.cloud.tools.eclipse.server.id";
+  public static final String SERVER_PROGRAM_FLAGS = Activator.PLUGIN_ID + ".SERVER_FLAGS";
+  public static final String DEFAULT_HOST = "localhost";
+  public static final int DEFAULT_API_PORT = 8188;
+  public static final int DEFAULT_ADMIN_PORT = 8000;
+
+  private static final String ATTR_CLOUD_SDK_SERVER_MODULES = "cloudsdk-server-modules-list";
+
+  private ILaunchConfigurationWorkingCopy remoteDebugLaunchConfig;
+
+  /**
+   * Returns a {@link CloudSdkServer} instance associated with the
+   * {@code server} or a new {@link CloudSdkServer} instance if a
+   * {@link CloudSdkServer} instance cannot be found for {@code server}.
+   *
+   * @param server the generic sever
+   * @return a new {@link CloudSdkServer} instance or the one associated with
+   *         {@code server}
+   */
+  public static CloudSdkServer getCloudSdkServer(IServer server) {
+    CloudSdkServer cloudSdkServer = server.getAdapter(CloudSdkServer.class);
+    if (cloudSdkServer == null) {
+      cloudSdkServer = (CloudSdkServer) server.loadAdapter(CloudSdkServer.class,
+                                                           new NullProgressMonitor());
+    }
+    return cloudSdkServer;
+  }
+
+  @Override
+  public IStatus canModifyModules(IModule[] add, IModule[] remove) {
+    if (add != null) {
+      for (IModule module : add) {
+        if (module.getProject() != null) {
+          IStatus status = FacetUtil.verifyFacets(module.getProject(), getServer());
+          if (status != null && !status.isOK()) {
+            return status;
+          }
+        }
+      }
+    }
+    return Status.OK_STATUS;
+  }
+
+  @Override
+  public IModule[] getChildModules(IModule[] module) {
+    if (module[0] != null && module[0].getModuleType() != null) {
+      IModule thisModule = module[module.length - 1];
+      IModuleType moduleType = thisModule.getModuleType();
+      if (moduleType != null && "jst.ear".equals(moduleType.getId())) { //$NON-NLS-1$
```

Comment:
Keep misspellings in one place ;)  More seriously, one place to check and it makes validations like your plugin.xml unit tests much more meaningful.

We can also give these more meaningful names. "jst.web" doesn't scream out "Servlet API" to me.